### PR TITLE
[BUGFIX] GE DataContext with multiple DataSources should not scan all configurations

### LINKED ISSUE #2241: DataContext scans through all datasources even if only one is being used
**Describe the bug**
When creating an Expectation suite with a datacontext that contains multiple datasources, GE will create connections to all datasources, even if only one is being used. 

**To Reproduce**

1. The following GE DataContext contains 2 Datasources that are connections to databases: one to a locally run postgres database (`my_postgres_db`) and one to a Redshift database (`my_redshift_db`). The postgres database is not actively running, but the configuration still exists.

`Select a datasource
    1. ge_data__dir
    2. my_postgres_db
    3. my_redshift_db
: 3`

2. Create an Expectation Suite and open the notebook to Edit Expectation Suite. 

Screen Shot 2021-01-07 at 10 49 11 PM

* Notice that `batch_kwargs` only contains a reference to `my_redshift_db`. 

3.  When the above code is run we get the following error about `my_postgres_db`

`DatasourceInitializationError: Cannot initialize datasource my_postgres_db, error: Cannot initialize datasource my_postgres_db, error: (psycopg2.OperationalError) could not connect to server: Connection refused
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 5433?
could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5433?

(Background on this error at: http://sqlalche.me/e/13/e3q8)
`

**Expected behavior**
`context.get_batch()` should only initialize the `my_redshift_db` datasource and ignore the connection to postgres since it is not configured in `batch_kwargs`

**Environment (please complete the following information):**
 - OS: Mac OS 11.0.1
 - GE Version: 0.13.4

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
Changes proposed in this pull request:
- In `DataContext`, the `datasource` property is now initialized with datasources that 1) are configured 2) GE currently has access to. These datasources are initialized and cached when `DataContext` is initialized.
- Previously accessing the `datasource` property resulted in the GE looping through all datasource configurations, which caused an error if any datasource configuration was 'stale' or no longer accessible. 
- closes #2241 
- Fixed a docstring
- Added unittest and fixture.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -2624,6 +2624,32 @@ def data_context_parameterized_expectation_suite(tmp_path_factory):
...
 
 
+@pytest.fixture
+def data_context_with_bad_datasource(tmp_path_factory):
+    """
+    This data_context is *manually* created to have the config we want, vs
+    created with DataContext.create()
+
+    This DataContext has a connection to a datasource named post_gres_db
+    which is not a valid datasource.
```
Comment (Shinnnyshinshin) [inline tests/conftest.py]:
```suggestion
    This DataContext has a connection to a datasource named my_postgres_db
    which is not a valid datasource.
```

File: great_expectations/data_context/types/base.py [hunk 2]
```diff
@@ -1206,9 +1206,6 @@ def __init__(
             commented_map = CommentedMap()
         self._commented_map = commented_map
         self._config_version = config_version
-        if datasources is None:
-            datasources = {}
-        self.datasources = datasources
```
Comment (Shinnnyshinshin) [inline great_expectations/data_context/types/base.py:1211]: `
if datasources is None:	
            datasources = {}	
        self.datasources = datasources
`

File: great_expectations/data_context/data_context.py [hunk 3]
```diff
@@ -1223,7 +1239,7 @@ def _get_data_context_version(self, arg1: Any, **kwargs) -> Optional[str]:
         Returns "v2" if the datasource is an instance of the LegacyDatasource class.
         """
 
-        if not self.datasources:
+        if self.datasources is {}:
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context.py]: The previous version `if not self.datasources` is a stronger check.

File: tests/test_fixtures/great_expectations_bad_datasource.yml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# This is a basic configuration for testing.
```
Comment (alexsherstinsky) [inline tests/test_fixtures/great_expectations_bad_datasource.yml:1]: We need this test for `config_version: 3` in addition to `config_version: 2`.  A pattern for doing this is now part of PR https://github.com/great-expectations/great_expectations/pull/2253 (against `checkpoint-class`).

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
1) A small change was requested (please see a comment).
2) I would strongly prefer for this PR to be against `checkpoint-class` and only after the PR https://github.com/great-expectations/great_expectations/pull/2253 is merged (hopefully no later than Thursday, 1/14/2021).  The Checkpoints works carries a number of enhancements in the areas of `Datasource` instantiation, the `config_version` change to 3 with the implications to tests, and in other areas -- and it is tracking `develop` closely.  Thank you!

File: great_expectations/data_context/data_context.py (hunk 3 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/data_context.py]: you're right. will change back :)

**Comment by Shinnnyshinshin** [comment]
Thanks for the review @alexsherstinsky.  I don't have a problem waiting a day, and merging to a different base branch. Let me know when #2253  is merged, and I'll ask for a re-review

**Comment by Shinnnyshinshin** [comment]
addressing conflicts now, and will ask for re-review

File: tests/data_context/test_data_context.py [hunk 5]
```diff
@@ -1908,3 +1909,34 @@ def test_get_validator_with_attach_expectation_suite(
         create_expectation_suite_with_name="A_expectation_suite",
     )
     assert my_validator.expectation_suite_name == "A_expectation_suite"
+
+
+def test_get_batch_multiple_datasources_do_not_scan_all(
+    data_context_with_bad_datasource,
+):
+    """
+    What does this test and why?
```
Comment (alexsherstinsky) [inline tests/data_context/test_data_context.py:1918]: @Shinnnyshinshin So basically, the stale/misconfigured data sources can exist, but they will not harm us; is this the enhancement?

**Review by alexsherstinsky — APPROVED** [review]
LGTM -- this is very nice and useful!  We might need to be expecting support calls if users expect their data sources to update on the fly, which we cannot do now.  Maybe we need a special PR, where one can request a refresh of the DataContext._cached_datasources property?  Curious as to what you think (if makes sense, then this can be added to Asana as a future enhancement idea).  Of course it may be just as easy for users to re-instantiate their DataContext -- not sure, as it might be use case depended.  Thank you very much!

**Comment by Shinnnyshinshin** [comment]
Hi @alexsherstinsky yes, you're right. If it's a helper method(s), then I think it can definitely be a useful enhancement, and hopefully pretty quick to implement :) Will add it to # flexipod enhancement list

**Pull request merged**